File: java/LinkedList.java
class Node {
  int data;
  Node next;

  public Node(int data) {
    this.data = data;
    this.next = null;
  }
}

public class LinkedList {
  private Node head;

  public LinkedList() {
    head = null;
  }
  // add a new node at beginning of list
  public void addFirst(int data) {
    Node newNode = new Node(data);
    newNode.next = head;
    head = newNode;
  }

// add new node at end of list 
public void addLast(int dat) {
  Node newNode = new Node(data);
  if(head == null) {
    head = newNode;
    return;
  }
  Node current = head;
  while(current.next != null) {
    current = current.next;
  }
  current.next = newNode;
}

// delete first node 
public void deleteFirst() {
  if(head == null) {
    return;
  }
  head = head.next;
}

// delete last node
public void deleteLast() {
  if(head == null || head.next == null) {
    head = null;
    return;
  }
  Node current = head;
  Node previous = null;
  while(current.next != null) {
    previous = current;
    current = current.next;
  }
  previous.next = null;
}

// print linkedlist
public void printList() {
  Node current = head;
  while(current != null) {
    System.out.print(current.data + " ");
    current = current.next;
  }
  System.out.println();
}
}

public class Main {
  public static void main(String[] args) {
    LinkedList list = new LinkedList)_;
    list.addLast(10);
    list.addLast(20);
    list.addFirst(5);
    list.printList();
    list.deleteFirst();
    list.printlist();
  }
}

File: javascript/linkedList.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
  }
  isEmpty() {
    return this.head === null;
  }
  addAtBeginning(data) {
    const newNode = new Node(data);
    newNode.next = this.head;
    this.head = newNode;
  }
  addAtEnd(data) {
    const newNode = new Node(data);
    if(this.isEmpty()) {
      this.head = newNode;
    } else {
      let current = this.head;
      while(current.next) {
        current = current.next;
      }
      current.next = newNode;
    }
  }
  deleteAtBeginning() {
    if(this.isEmpty()) {
      return;
    }
    this.head = this.head.next;
  }
  deleteAtEnd() {
    if(this.isEmpty()) {
      return;
    }
    if(this.head.next === null) {
      this.head = null;
    } else {
      let current = this.head;
      while (current.next.next) {
        current = current.next;
      }
      current.next = null;
    }
  }

  printList() {
    let current = this.head;
    while (current) {
      console.log(current.data);
      current = current.next;
    }
  }
}
